src/jdk.management/unix/classes/com/sun/management/internal/OperatingSystemImpl.java

```diff
@@ -42,8 +45,130 @@
 
     private static final int MAX_ATTEMPTS_NUMBER = 10;
     private final Metrics containerMetrics;
-    private long usageTicks = 0; // used for cpu load calculation
-    private long totalTicks = 0; // used for cpu load calculation
+    private ContainerCpuTicks systemLoadTicks = new SystemCpuTicks();
+    private ContainerCpuTicks processLoadTicks = new ProcessCpuTicks();
+
+    private abstract class ContainerCpuTicks {
+        private long usageTicks = 0;
+        private long totalTicks = 0;
+
+        private double getUsageDividesTotal(long usageTicks, long totalTicks) {
+            // If cpu quota or cpu shares are in effect. Calculate the cpu load
+            // based on the following formula (similar to how
+            // getCpuLoad0() is being calculated):
+            //
+            //   | usageTicks - usageTicks' |
+            //  ------------------------------
+            //   | totalTicks - totalTicks' |
+            //
+            // where usageTicks' and totalTicks' are historical values
+            // retrieved via an earlier call of this method.
+            if (usageTicks < 0 || totalTicks <= 0) {
+                return -1;
+            }
+            long distance = usageTicks - this.usageTicks;
+            this.usageTicks = usageTicks;
+            long totalDistance = totalTicks - this.totalTicks;
+            this.totalTicks = totalTicks;
+            double systemLoad = 0.0;
+            if (distance > 0 && totalDistance > 0) {
+                systemLoad = ((double)distance) / totalDistance;
+            }
+            // Ensure the return value is in the range 0.0 -> 1.0
+            systemLoad = Math.max(0.0, systemLoad);
+            systemLoad = Math.min(1.0, systemLoad);
+            return systemLoad;
+        }
+
+        public double getContainerCpuLoad() {
+            assert(containerMetrics != null);
+            long quota = containerMetrics.getCpuQuota();
+            long share = containerMetrics.getCpuShares();
+            if (quota > 0) {
+                long numPeriods = containerMetrics.getCpuNumPeriods();
+                long quotaNanos = TimeUnit.MICROSECONDS.toNanos(quota * numPeriods);
+                return getUsageDividesTotal(cpuUsageSupplier().getAsLong(), quotaNanos);
+            } else if (share > 0) {
+                long hostTicks = getHostTotalCpuTicks0();
+                int totalCPUs = getHostOnlineCpuCount0();
+                int containerCPUs = getAvailableProcessors();
+                // scale the total host load to the actual container cpus
+                hostTicks = hostTicks * containerCPUs / totalCPUs;
+                return getUsageDividesTotal(cpuUsageSupplier().getAsLong(), hostTicks);
+            } else {
```

Comment:
The old code had these comments here. We should add them as they're still relevant:

```
                // If CPU quotas and shares are not active then find the average system load for
                // all online CPUs that are allowed to run this container.

                // If the cpuset is the same as the host's one there is no need to iterate over each CPU
```